src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb

```diff
@@ -22625,6 +22625,134 @@ End Class
+End Interface
+
+Public Interface I1(Of T)
+    Inherits I2(Of T, (a As Object, b As Object))
+End Interface
+"
+            Dim lib1_comp = CreateCompilationWithMscorlib40(lib1_source, references:={getReference(valueTuple_comp), getReference(tupleElementNamesAttribute_comp)})
+            lib1_comp.AssertNoDiagnostics()
+
+            Dim lib2_source = "
+Public interface I0
+    Inherits I1(Of string)
+End Interface
+"
+            Dim lib2_comp = CreateCompilationWithMscorlib40(lib2_source, references:={getReference(lib1_comp), getReference(valueTuple_comp)}) ' Missing TupleElementNamesAttribute
+            lib2_comp.AssertNoDiagnostics()
+            lib2_comp.AssertTheseEmitDiagnostics()
+
+            Dim imc1 = CType(lib2_comp.GlobalNamespace.GetMember("I0"), TypeSymbol)
+            AssertEx.SetEqual({"I1(Of System.String)"}, imc1.InterfacesNoUseSiteDiagnostics().Select(Function(i) i.ToTestDisplayString()))
+            AssertEx.SetEqual({"I1(Of System.String)", "I2(Of System.String, (a As System.Object, b As System.Object))"}, imc1.AllInterfacesNoUseSiteDiagnostics.Select(Function(i) i.ToTestDisplayString()))
+
+            Dim client_source = "
+Public Class C
+    Public Sub M(imc As I0)
+        imc.ExecuteAsync("""")
+    End Sub
+End Class
+"
+            Dim client_comp = CreateCompilationWithMscorlib40(client_source, references:={getReference(lib1_comp), getReference(lib2_comp), getReference(valueTuple_comp)})
+            client_comp.AssertNoDiagnostics()
+
+            Dim imc2 = CType(client_comp.GlobalNamespace.GetMember("I0"), TypeSymbol)
+            AssertEx.SetEqual({"I1(Of System.String)"}, imc2.InterfacesNoUseSiteDiagnostics().Select(Function(i) i.ToTestDisplayString()))
+            AssertEx.SetEqual({"I1(Of System.String)", "I2(Of System.String, (a As System.Object, b As System.Object))"}, imc2.AllInterfacesNoUseSiteDiagnostics.Select(Function(i) i.ToTestDisplayString()))
+
+        End Sub
+
+        <Fact, WorkItem(40033, "https://github.com/dotnet/roslyn/issues/40033")>
+        Public Sub SynthesizeTupleElementNamesAttributeBasedOnInterfacesToEmit_BaseAndDirectInterface()
+
+            Dim source = "
+Namespace System
+    Public Structure ValueTuple(Of T1, T2)
+        Public Dim Item1 As T1
+        Public Dim Item2 As T2
+
+        Public Sub New(item1 As T1, item2 As T2)
+            me.Item1 = item1
+            me.Item2 = item2
+        End Sub
+
+        Public Overrides Function ToString() As String
+            Return ""{"" + Item1?.ToString() + "", "" + Item2?.ToString() + ""}""
+        End Function
+    End Structure
+End Namespace
+
+Namespace System.Runtime.CompilerServices
+    Public Class TupleElementNamesAttribute
+        Inherits Attribute
+
+        Public Sub New() ' Note: bad signature
+	    End Sub
+    End Class
+End Namespace
+
+Public Interface I(Of T)
+End Interface
+
+Public Class Base(Of T)
+End Class
+
+Public Class C1
+    Implements I(Of (a As Object, b As Object))
```

Comment:
>I(Of (a As Object, b As Object)) [](start = 15, length = 32)

@AlekseyTs This is checked during binding (both in C# and VB) since the tuple appears in syntax.
Here's the check in VB: [https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Portable/Binding/Binder_Symbols.vb#L665](https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Portable/Binding/Binder_Symbols.vb#L665)

This means that the only missing check for tuple names attribute for C# was indirect interfaces.
We're currently reporting a duplicate diagnostic in C# in latest commit (iteration 9). I don't see a big benefit to removing that duplicate, as this seems a rare scenario (System.ValueTuple and TupleElementNamesAttribute both typically live in the same reference).